### ISSUE TITLE
BUG: Use of union types breaks proxy building, creates broken PHP code

### ISSUE THREAD

**Issue by mficzel** [neos/flow-development-collection#2941]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When union types are used as return values in methods that are proxied types like `int|false` are interpreted as `\int|false` which leads to php errors.

### Expected Behavior

Union types should be allowed to be used as method arguments and return values also on class-methods that are proxied. At the very least an error should be raised and notify the user about the missing support for union types.

### Steps To Reproduce

Crete a class with a method that uses union types and somehow force flow to crate an proxy for that. An aspect on that method should do the trick. You should now get a php error.

### Environment

```markdown
- Flow: 7.3
- PHP: 8.0
```

### Anything else?

Union types are allowed since php 8 which is the minimal php version flow currently supports. Given this it can be expected that union types can be used in code based on flow.

**Comment by kdambekalns** [comment]
Nice issue title! 😜

**Comment by albe** [comment]
There is *some* support for union types as of Flow 7.2 (https://github.com/neos/flow-development-collection/pull/2404) but might very well be there are still some cases not covered.

First thing to do would be to create a failing test case in
https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php
and adjust the fixture https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Tests/Functional/ObjectManagement/Fixtures/PHP8/ClassWithUnionTypes.php

**Comment by mficzel** [comment]
@albe flushed all caches again and still get the wrong return type. I made a draft pr with the branch i found this on https://github.com/neos/flow-development-collection/pull/2943 ... problem occurs in the Session:collectGarbage(): int|false and is visible in the functional tests aswell. After removing the redundant / in proxy class manually all tests are running so i am pretty sure this is caused by the union type.

The tests you mentioned mainly ensure that `stuff|null` is handled like `?stuff` which is not using `ReflectionUnionType` but `ReflectionNamedType`. It also ensures that ReflectionUnionType is returned on the other Types which seems to work fine. However i suspect the error in the actual proxy building and not the reflection.

**Comment by albe** [comment]
> The tests you mentioned mainly ensure that stuff|null is handled like ?stuff

Ah yes. I knew there was something merely partial in all that 🙈 

> However i suspect the error in the actual proxy building and not the reflection.

I don't remember the escaping of return types happening in the proxy builder, but I might have forgotten something.
If I'm not, then the line causing the issue might be https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Classes/Reflection/ReflectionService.php#L1451-L1453 - I suppose we would need to `split` a potential union type and escape every single one, then join back together with `|`

**Comment by mficzel** [comment]
@albe you are right the line you mentioned is at least part of the problem. Added a test in #2944 ...

**Comment by robertlemke** [comment]
Flow also creates proxy classes with invalid code if a method signature contains arguments with union types and they are protected via method privileges in a Flow policy.

For example, protecting this method in a policy will break:

```php
    public function get(DomainName|Sha1 $domainNameOrSha1): Domain
    {…}
```

This results in the following error (PHP namespaces removed for brevity):

> Fatal error: Could not check compatibility between …DomainService::get(…\DomainName|…\Sha1 $domainNameOrSha1): …\Domain and …DomainService_Original::get(…\DomainName|…\Sha1 $domainNameOrSha1): …\Domain, because class …\Sha1 is not available in …/Cache/Code/Flow_Object_Classes/…_DomainService.php on line 206

**Comment by robertlemke** [comment]
I assign this issue to myself and plan to work on it during the next sprint. If anyone has time to tackle this before me, feel free to assign it to yourself instead.

**Comment by robertlemke** [comment]
This should now be fixed in 9.0, see https://github.com/neos/flow-development-collection/pull/3057
